libpod/image/image.go

```diff
@@ -1075,3 +1077,62 @@ func (i *Image) Comment(ctx context.Context, manifestType string) (string, error
 	}
 	return ociv1Img.History[0].Comment, nil
 }
+
+// Tree gets dependencies of image
+func (i *Image) Tree(ctx context.Context) (*bytes.Buffer, error) {
+	treeDepth := 1
+	rootImage := i
+
+	//Traverse to the root(base) image
+	parent, err := i.GetParent()
+	if err != nil {
+		return nil, err
+	}
+
+	for parent != nil {
+		rootImage = parent
+		parent, err = parent.GetParent()
```

Comment:
This seems problematic; the more complex code, or maybe (more likely?) an entirely different data source might still be needed.

---

Most importantly, AFAICS there may not be image objects available at all for the shared parents: consider a _layer_ tree like this:

- `fedora`
  - unnamed layer
    - unnamed layer
    - `base-environment`
      - `app-0`
      - unnamed layer
        - unnamed layer [A]
          - `app-1`
      - unnamed layer
        - unnamed layer
          - `app-2`

and a local environment which has done `for x in 0 1 2; do podman pull app-$x; done`, with no other images pulled or known, so that the `fedora` and `base-environment` images have never been seen locally, nor have any possible in-between images created by `buildah`. 

AFAICS (but I have not tested this, to be fair), in such case the tree of layers in the `containers/storage.LayerStore` will mirror the structure above exactly, but _only_ the `app-$x` images will be known to `containers/storage.ImageStore`.

So, `app-1`.`GetParent()` will return `nil` because the layer `[A]` has no corresponding image; similarly `app-0`.`GetParent()` will return `nil` because the `base-environment` image is not available locally.

In both cases, the tree will only show a single element `app-$0`. It seems to me (but, strictly speaking, I don’t know, because the specification is not yet precise enough) that the desired input is the full layer tree exactly as structured, with `fedora` and `base-environment` reported as “unnamed layer” [but after `podman pull base-environment`, it could report the right image name].

---

Second, as discussed in 
https://github.com/containers/libpod/issues/415#issuecomment-429135936 , the incomplete data may be ambiguous: there might be several images with the same top layer but different configs, and `GetParent` chooses one in random order (AFAICS the first one pulled/created in the local storage, but that’s not an API promise).

In this case, consider an image `base-modified`, derived from `base-environment` with an `ENV a=a`. Depending on the implementation, this might cause an empty extra layer to be created, but there’s no real need for it: it’s just a different image with a different config but exactly the same layers. So, the `base-environment` layer actually can have both `base-environment` and `base-modified` images as possible parents, and without knowing the true parent links somehow (AFAICS `buildah`/`podman` does not record that at all right now), there’s no way to tell.

If both are available locally, we can’t tell which one to output; if only _the wrong one_ is available locally, we will probably report that incorrect base. (And, as discussed above, neither may be available locally, and we can’t tell between an unavailable base image and a non-image intermediate layer.)

AFAICS this is fundamentally unsolvable in general, but it is a design concern — do we want to report a single, possibly incorrect, image, or all of the candidates, of which all but one are certainly incorrect, and the last one may _or may not_ be correct?)

Eventually, we have been talking about adding true/reliable “parent image” information, which would make this unambiguous — _only if_ the image is new enough to have been built with such hypothetical parent information; we would still have to guess for any image that exists today.

Reply:
>So, app-1.GetParent() will return nil because the layer [A] has no corresponding image; similarly app-0.GetParent() will return nil because the base-environment image is not available locally.

No, In this case a layerID (_maybe without RepoTags_) will be shown if image is not flatten/squashed by any tool. 
   - If image is flatten/squashed, it has become independent, so no there will be no parent layer.
While pulling children layer, always dependent parent layers are downloaded and that’s how the image rootfs should be build.
For build information `podman history` should be used

> with fedora and base-environment reported as “unnamed layer” [but after podman pull base-environment, it could report the right image name].
 
AFAIU `tree` or any other script that finds dependencies locally, resolves to local registry. i.e. information available locally instead of fetching to remote registry/store.

If it involves fetching information from remote registry, than we need to decide which all remote registries need to be looked at also, as podman supports multiple registries.

Reply:
> No, In this case a layerID (maybe without RepoTags) will be shown if image is not flatten/squashed by any tool.

How does that work? https://github.com/containers/libpod/blob/604728d65dd0110b28c79acafc5b214886b3b484/libpod/image/image.go#L984-L999 clearly _can_ return _nil_ if there is no image corresponding to a layer, and reading `c/storage/layerStore.Put`, creating a layer does not require a corresponding image to exist.

> If it involves fetching information from remote registry, than we need to decide which all remote registries need to be looked at also, as podman supports multiple registries.

That’s the least of our problems; there is no way to look up a remote image config by the layer set (and again, this can be return many different results with no way to decide), and no way to lookup a remote manifest by the config.